### ISSUE TITLE
/dbs/<dbname>/_changes memory leak

### ISSUE THREAD

**Issue by bjendyk** [pouchdb/express-pouchdb#398]
When tested with apache bench tool using the following command:

ab -k -c 10 -n 10000 http://localhost:port/dbs/<db_name>/_changes

the heap usage grows continuously over time. I have included **memwatch-next** module to record some memory usage statistics. During this test (10000 requests), the heap size ("current_base" as reported by memwatch in "stats" event) starts with something around 30 MB, and ends somewhere around 130 MB after 154 garbage collection operations. I'm including the captured values (X being the consecutive number of garbage collection, Y being the current heap size) as a CSV to plot them and observe the leak as a reference.

express-pouchdb version: 2.1.0
node.js version: 5.10.1
express version: 4.14.0
pouchdb version: 6.0.5

See https://github.com/bjendyk/express-pouchdb-memleak app that helps to reproduce the problem.

Stats recorded by memwatch:

num_of_gc, current_base
1,31206792
2,28568336
3,38428640
4,25790680
5,42686912
6,49511040
7,28185576
8,51364960
9,29323784
10,52470064
11,54629264
12,31164792
13,57209752
14,32415744
15,57673592
16,55494728
17,33849592
18,50667872
19,54537696
20,56950304
21,58198880
22,63423744
23,39145240
24,61749440
25,40337888
26,56256088
27,39856192
28,63509208
29,40758576
30,60000352
31,42044688
32,61730920
33,43065792
34,62966336
35,69602512
36,75138584
37,76913808
38,78103848
39,46804392
40,60956352
41,67628552
42,48323568
43,64079808
44,71439592
45,79203792
46,52024488
47,69887096
48,53387112
49,72078336
50,54417904
51,73670072
52,55198360
53,75348632
54,82796680
55,56299664
56,71773920
57,87209752
58,58834576
59,75504936
60,59225656
61,77072208
62,78729952
63,60899688
64,80997416
65,61999920
66,81186184
67,61822824
68,83539840
69,63260584
70,83565456
71,85331720
72,64219744
73,85082672
74,65488120
75,90123952
76,66603496
77,91973000
78,67410384
79,93429616
80,93942656
81,97138632
82,70628056
83,94738808
84,70173216
85,97763072
86,94758744
87,72641152
88,88965088
89,72597432
90,91951040
91,74447192
92,92560376
93,74609264
94,91009024
95,102412344
96,101097824
97,76836520
98,98351528
99,78205880
100,95485368
101,78078336
102,96942152
103,79611424
104,97966560
105,80352192
106,102432664
107,107000320
108,81330560
109,101571104
110,83077800
111,102097000
112,82995576
113,103731632
114,83949008
115,104518776
116,112659760
117,85658232
118,107206256
119,87117720
120,108367176
121,116743040
122,89202736
123,110789448
124,89620464
125,111552504
126,111816456
127,91092480
128,117858328
129,91777768
130,119158232
131,93611328
132,120514800
133,94432648
134,113516248
135,95165552
136,114461784
137,96148096
138,116186720
139,120682808
140,98270288
141,127296720
142,99196064
143,128064008
144,98996344
145,121097184
146,100856000
147,130779880
148,101012776
149,124165968
150,103021224
151,125060192
152,103308584
153,126713384
154,134001336

**Comment by gr2m** [comment]
I’ve found out today that [lib/routes/changes.js#L30-L33](https://github.com/pouchdb/express-pouchdb/blob/67f7b26a16f64add7cc7030f598c12485b07810d/lib/routes/changes.js#L30-L33) starts the interval for the heartbeat but doesn’t always end it. For me it manifested by causing timeouts in my e2e tests, because the process would not end and eventually time out. [wtfnode](http://npm.im/wtfnode) helped find the cause eventually.

can you check if my PR here helps: https://github.com/pouchdb/express-pouchdb/pull/403

**Comment by bjendyk** [comment]
It seems that the problem is not fixed. I updated to 2.1.1 and ran https://github.com/bjendyk/express-pouchdb-memleak. The heap size growth is even bigger than using 2.1.0.

**Comment by gr2m** [comment]
@bjendyk can you investigate what it might be? I don’t know much about profiling, but if you can give me some hints I can try help to fix the problem

**Comment by sphaero** [comment]
I can confirm the memleak. Just tried the test app as well with around 10k requests.

```
1,27000248
2,29879728
3,37309096
4,51136408
5,54990272
6,65065440
7,64944384
8,71602880
9,78879168
10,86713424
11,93985048
12,101098136
13,108545072
14,116531984
```

**Comment by nolanlawson** [comment]
Thank you for the solid repro. We should really look at this.

**Comment by nolanlawson** [comment]
This is incredibly easy to repro. You can repro it with a simple:

    curl -sL -X PUT localhost:6984/testdb
    for ((i=0;i< 1000;i++)); do curl -sL localhost:6984/test/_changes; done

Using the `--inspect` flag and the Memory tool in the Dev Tools, you can see the memory increasing. It seems related to the fact that we call `new PouchDB()` for each call, which is probably also something that can be optimized since it's a bit wasteful.

**Comment by nolanlawson** [comment]
Fixed in https://github.com/pouchdb/express-pouchdb/pull/426 and published in express-pouchdb@2.2.2
